service/hook/assembla/assembla.go

```diff
@@ -0,0 +1,140 @@
+package assembla
+
+//
+// Docs: https://articles.assembla.com/assembla-basics/learn-more/post-information-to-external-systems-using-webhooks
+//
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/bitrise-io/bitrise-webhooks/bitriseapi"
+	hookCommon "github.com/bitrise-io/bitrise-webhooks/service/hook/common"
+)
+
+// --------------------------
+// --- Webhook Data Model ---
+
+// SpaceEventModel ...
+type SpaceEventModel struct {
+	Space	string	`json:"space"`
+	Action	string	`json:"action"`
+	Object	string	`json:"object"`
+}
+
+// MessageEventModel ...
+type MessageEventModel struct {
+	Title   string	`json:"title"`
+	Body    string	`json:"body"`
+	Author	string	`json:"author"`
+}
+
+// GitEventModel ...
+type GitEventModel struct {
+	RepositorySuffix	string      `json:"repository_suffix"`
+	RepositoryURL		string      `json:"repository_url"`
+	Branch				string      `json:"branch"`
+	CommitID			string      `json:"commit_id"`
+}
+
+// PushEventModel ...
+type PushEventModel struct {
+	SpaceEventModel	    SpaceEventModel		`json:"assembla"`
+	MessageEventModel	MessageEventModel	`json:"message"`
+	GitEventModel		GitEventModel		`json:"git"`
+}
+
+// ---------------------------------------
+// --- Webhook Provider Implementation ---
+
+// HookProvider ...
+type HookProvider struct{}
+
+func detectContentType(header http.Header) (string, error) {
+	contentType := header.Get("Content-Type")
+	if contentType == "" {
+		return "", errors.New("No Content-Type Header found")
+	}
+
+	return contentType, nil
+}
+
+func detectAssemblaData(pushEvent PushEventModel) error {
+	if (pushEvent.GitEventModel.CommitID == "---") ||
+		(pushEvent.GitEventModel.Branch == "---") ||
+		(pushEvent.GitEventModel.RepositoryURL == "---") ||
+		(pushEvent.GitEventModel.RepositorySuffix == "---")	{
+		return errors.New("Webhook is not correctly setup, make sure you post updates about 'Code commits' in Assembla")
+	}
+
+	return nil
+}
+
+func transformPushEvent(pushEvent PushEventModel) hookCommon.TransformResultModel {
+	if pushEvent.SpaceEventModel.Action != "committed" {
+		return hookCommon.TransformResultModel{
+			Error: fmt.Errorf("Action was not 'committed', was: %s", pushEvent.SpaceEventModel.Action),
+		}
+	}
+	if pushEvent.MessageEventModel.Body == "" {
+		return hookCommon.TransformResultModel{
+			Error: fmt.Errorf("Message body can't be empty"),
+		}
+	}
+	if pushEvent.MessageEventModel.Author == "" {
+		return hookCommon.TransformResultModel{
+			Error: fmt.Errorf("Message author can't be empty"),
+		}
+	}
+	if pushEvent.GitEventModel.Branch == "" {
+		return hookCommon.TransformResultModel{
+			Error: fmt.Errorf("Git branch can't be empty"),
+		}
+	}
+
+	triggerAPIParams := []bitriseapi.TriggerAPIParamsModel{
+		{
+			BuildParams: bitriseapi.BuildParamsModel{
+				CommitMessage: pushEvent.MessageEventModel.Body,
+				Branch:        pushEvent.GitEventModel.Branch,
```

Comment:
Use `CommitID` to specify the commit hash! E.g.: https://github.com/bitrise-io/bitrise-webhooks/blob/master/service/hook/github/github.go#L114

Without that the build will build the tip of the branch, and a rebuild of the build will also build the tip of the branch (instead of the same commit hash)!

This means that e.g. if you push a commit to the `master` branch, that starts a build on bitrise.io, then you push more code to `master` and then you try to rebuild a previous build that won't build the same commit hash, that will build the latest version of `master`. The intended behaviour of these webhook integrations is to include all the required infos for a rebuild, so that a rebuild actually re-builds the exact same state of the repository.